### PR TITLE
Allow unbuffered output from Ansible/Python

### PR DESCRIPTION
**Pull request by deployable**
https://github.com/shaharke/node-ansible/issues/11

### CODE REVIEW

File: lib/ansible.js [hunk 1]
```diff
@@ -24,11 +25,20 @@ AbstractAnsibleCommand.prototype.exec = function(options) {
...
   }
 
+  // Make Ansible output unbuffered so the stdout and stderr `data` events 
+  // pickup output up more regularly
+  var unbuffered = (options.buffered) ? '' : 1;
+  processOptions.env = process.env;
+  _.extend( processOptions.env, { PYTHONUNBUFFERED: unbuffered } );
+
   var output = '';
+  debug('exec',this.commandName(), this.compileParams(), processOptions);
```
Comment (shaharke) [inline lib/ansible.js]: Yeah... I didn't quite buy into this library. In general I want to reduce verbosity by the library or at least allow uses to pass their own logger.

Comment (deployable) [inline lib/ansible.js]: Ah.. sorry, I was using that for my testing. It wasn't meant to make it into the commit.

Comment (deployable) [inline lib/ansible.js]: as a side node, the lack of verbosity, or control of it, is the main feature I like about the debug library. The `debug` call is a noop unless you specifically ask for that debug string/class to be turned on so it works really well for multi file/class apps, when you setup the naming structure.  It's always a nice surprise when you turn DEBUG='*' on and your dependencies start doing debug as well.

File: test/playbook.spec.js [hunk 2]
```diff
@@ -144,9 +147,36 @@ describe('Playbook command', function () {
...
+      var command = new Playbook().playbook('test');
+      var promise = command.exec();
+      expect(promise).to.be.fulfilled.then(function () {
+        expect(spawnSpy).to.be.calledWith('ansible-playbook', ['test.yml']);
+        expect(spawnSpy.getCall(0).args[2]).to.have.deep.property( "env.PYTHONUNBUFFERED", "1" );
+        done();
+      }).done();
+    })
+
+    it('should turn on buffering when told too', function (done) {
```
Comment (shaharke) [inline test/playbook.spec.js]: told to :)

File: test/playbook.buffering.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
+// test for buffered and unbuffered output
```
Comment (shaharke) [inline test/playbook.buffering.js]: Maybe put this in the examples dir instead?

File: lib/ansible.js [hunk 4]
```diff
@@ -24,10 +24,18 @@ AbstractAnsibleCommand.prototype.exec = function(options) {
...
+  if (!options) options = {};
+
+  if (options.cwd) {
     processOptions.cwd = options.cwd;
   }
 
+  // Make Ansible output unbuffered so the stdout and stderr `data` events 
+  // pickup output up more regularly
+  var unbuffered = (options.buffered) ? '' : 1;
+  processOptions.env = process.env;
```
Comment (shaharke) [inline lib/ansible.js]: Does it really makes sense to use the parent process env in this case?

File: lib/ansible.js (hunk 1 above)
Comment (shaharke) [inline lib/ansible.js]: I don't like the fact that the output target for the debug messages is not under the control of the caller. In a bit more complex systems it complicates log management.

File: lib/ansible.js (hunk 4 above)
Comment (deployable) [inline lib/ansible.js]: That's what spawn [defaults to](https://github.com/nodejs/node/blob/c22bcd31f082747522dad355f987845cd7b9f5e8/lib/child_process.js#L314) when nothing is supplied.

Comment (shaharke) [inline lib/ansible.js]: Yeah. Your code just made we think that this has been a mistake on my part, and in fact we need to reset the environment. Not critical though for this PR.

File: lib/ansible.js (hunk 1 above)
Comment (deployable) [inline lib/ansible.js]: Cool. I guess that's what you sign up for by using it, just stderr debug.

File: test/playbook.spec.js (hunk 2 above)
Comment (deployable) [inline test/playbook.spec.js]: :blush:

File: test/playbook.buffering.js (hunk 3 above)
Comment (deployable) [inline test/playbook.buffering.js]: Done. I guess I was looking for a separate `functional` test suite for mocha to launch outside of the normal `npm test` suite that actually launches ansible.

File: lib/ansible.js (hunk 4 above)
Comment (deployable) [inline lib/ansible.js]: I was thinking about that when I hit the issue, It was really only `PATH` that was needed to launch `ansible-playbook` but I went with the whole environment as I'm sure there could be other vars that ansible might rely on. 

For example when initially when testing this, I expected that `PYTHONUNBUFFERED` in the shell launching `node` would be passed through. You could build up that list of vars to be passed through but things will break in the mean time.

Comment (shaharke) [inline lib/ansible.js]: Ok. Let's keep it that way.

**Comment by shaharke** [comment]
@deployable published

**Pull request merged**